src/main/scala/zio/prelude/NonEmptySet.scala

```diff
@@ -0,0 +1,181 @@
+package zio.prelude
+
+import zio.NonEmptyChunk
+
+import scala.language.implicitConversions
+
+final class NonEmptySet[A] private (private val set: Set[A]) { self =>
+
+  /** Converts this `NonEmptySet` to a `Set`. */
+  def toSet: Set[A] = set
+
+  /**
+   * Returns an element of this `NonEmptySet` and the remainder, which is a (possibly empty) `Set`.
+   */
+  @inline
+  def destruct: (A, Set[A]) = (set.head, set.tail)
+
+  /**
+   * Converts this `NonEmptySet` to a `NonEmptyChunk`.
+   */
+  def toNonEmptyChunk: NonEmptyChunk[A] = destruct match { case (head, tail) => NonEmptyChunk.fromIterable(head, tail) }
+
+  /**
+   * Converts this `NonEmptySet` to a `NonEmptyList`.
+   */
+  def toNonEmptyList: NonEmptyList[A] = destruct match { case (head, tail) => NonEmptyList.fromIterable(head, tail) }
+
+  /** Creates a new `NonEmptySet` with an additional element, unless the element is
+   *  already present.
+   *
+   *  @param elem the element to be added
+   *  @return a new set that contains all elements of this set and that also
+   *          contains `elem`.
+   */
+  def +(elem: A): NonEmptySet[A] = new NonEmptySet(set + elem)
+
+  /** Computes the union between of `NonEmptySet` and another set.
+   *
+   *  @param   that  the set to form the union with.
+   *  @return  a new `NonEmptySet` consisting of all elements that are in this
+   *  set or in the given set `that`.
+   */
+  def union(that: Set[A]): NonEmptySet[A] = new NonEmptySet(set.union(that))
+
+  /** Creates a new `NonEmptySet` by adding all elements contained in another collection to this `NonEmptySet`, omitting duplicates.
+   *
+   * This method takes a collection of elements and adds all elements, omitting duplicates, into `NonEmptySet`.
+   *
+   * Example:
+   *  {{{
+   *    scala> val a = NonEmptySet(1, 2) ++ NonEmptySet(2, "a")
+   *    a: zio.prelude.NonEmptySet[Any] = NonEmptySet(1, 2, a)
+   *  }}}
+   *
+   *  @param elems     the collection containing the elements to add.
+   *  @return a new `NonEmptySet` with the given elements added, omitting duplicates.
+   */
+  def ++(elems: Iterable[A]): NonEmptySet[A] = new NonEmptySet(set ++ elems)
+
+  /** Adds the `elem` to this `NonEmptySet`. Alias for `+`. */
+  def add(elem: A): NonEmptySet[A] = self + elem
+
+  /** Removes the `elem` from this `NonEmptySet`. Alias for `-`. */
+  def remove(elem: A): Set[A] = set - elem
+
+  /**
+   * Flattens a `NonEmptySet` of `NonEmptySet` values into a single
+   * `NonEmptySet`.
+   */
+  def flatten[B](implicit ev: A <:< NonEmptySet[B]): NonEmptySet[B] =
+    new NonEmptySet[B](set.foldLeft[Set[B]](Set.empty)((b, a) => b union ev(a)))
+
+  private lazy val hashCodeLazy: Int = set.hashCode ^ getClass.hashCode
```

Comment:
Sorry, didn't mean we had to cache the whole hash code but just the seed for the non-empty set. Something like:

```scala
class NonEmptySet {
  override def hashCode: Int = set.hashCode ^ nonEmptySetSeed
}

object NonEmptySet {
  private val NonEmptySetSeed: Int = 1247820194
```

Reply:
But isn't that already better? The whole hash is cached and we don't need to come up with any seed ourselves.

Reply:
Well I think using a specific number versus `getClass.hashCode` is just a small optimization so we don't have to do that additional operation. Then a second thing is that portion of the hashCode only needs to be computed once for the entire application. Even if we cache the hash code of each individual entry we end up computing that portion once for each instance instead of only once for the entire application. So those seem like clear wins.

Caching the whole hash code for each individual instance seems like more of a mixed bag since there is also a cost to storing these things and using lazy values. So you definitely see caching hash codes sometimes for data types where it is expensive to hash or is going to be used often but most of the time I don't think you cache the whole hash code for generic data types like this.